third_party/nvidia/backend/driver.py

```diff
@@ -159,254 +159,42 @@ def format_of(ty):
-  Py_END_ALLOW_THREADS;
-  if (PyErr_Occurred()) {{
-    return NULL;
-  }}
-
-  if(launch_exit_hook != Py_None){{
-    PyObject* args = Py_BuildValue("(O)", launch_metadata);
-    PyObject* ret = PyObject_CallObject(launch_exit_hook, args);
-    Py_DECREF(args);
-    if (!ret)
-      return NULL;
-
-  }}
-
-  // return None
-  Py_INCREF(Py_None);
-  return Py_None;
-}}
-
-static PyMethodDef ModuleMethods[] = {{
-  {{"launch", launch, METH_VARARGS, "Entry point for all kernels with this signature"}},
-  {{NULL, NULL, 0, NULL}} // sentinel
-}};
-
-static struct PyModuleDef ModuleDef = {{
-  PyModuleDef_HEAD_INIT,
-  \"__triton_launcher\",
-  NULL, //documentation
-  -1, //size
-  ModuleMethods
-}};
-
-PyMODINIT_FUNC PyInit___triton_launcher(void) {{
-  PyObject *m = PyModule_Create(&ModuleDef);
-  if(m == NULL) {{
-    return NULL;
-  }}
-  PyModule_AddFunctions(m, ModuleMethods);
-  return m;
-}}
-"""
+
+    def gen_c_def_macro(macro_name, macro_value):
+        return f"#define {macro_name} {macro_value}\n"
+
+    # macros to define:
+    """
+    #define EXTRA_INNER_LAUNCH_PARAM_DECLS
+    #define INNER_LAUNCH_CUDA_CHECK_ARGS
+    #define LAUNCH_PY_ARGS
+    #define PY_ARG_FORMAT_STR
+    #define EXTRA_LAUNCH_PARSE_PY_ARGS
+    #define DEVICE_PTR_INFO_VARS
+    #define TMA_DESC_VARS
+    #define EXTRA_INNER_LAUNCH_CALL_ARGS
+    """
+    macro_defs = gen_c_def_macro("EXTRA_INNER_LAUNCH_PARAM_DECLS", ", " + arg_decls if arg_decls else "")
+    macro_defs += gen_c_def_macro("INNER_LAUNCH_CUDA_CHECK_ARGS", ', '.join(f"&arg{i}" for i in params))
+    macro_defs += gen_c_def_macro("LAUNCH_PY_ARGS",
+                                  ';'.join([f"{_extracted_type(ty)} _arg{i}" for i, ty in signature.items()]))
+    macro_defs += gen_c_def_macro("PY_ARG_FORMAT_STR", f'"{format}"')
+    macro_defs += gen_c_def_macro("EXTRA_LAUNCH_PARSE_PY_ARGS", ", " + args_list if args_list else "")
+    device_ptr_info_var_list = []
+    tma_desc_var_list = []
+    for i, ty in signature.items():
+        if ty[0] == "*":
+            device_ptr_info_var_list.append(
+                f"DevicePtrInfo ptr_info{i} = getPointer(_arg{i}, {i}); if (!ptr_info{i}.valid) return NULL;")
+        elif ty == "nvTmaDesc":
+            tma_desc_var_list.append(f"CUtensorMap* tma_ptr{i} = getTmaDesc(_arg{i}); if (!tma_ptr{i}) return NULL;")
+
+    macro_defs += gen_c_def_macro("DEVICE_PTR_INFO_VARS", "  \\\n".join(device_ptr_info_var_list))
+    macro_defs += gen_c_def_macro("TMA_DESC_VARS", "  \\\n".join(tma_desc_var_list))
+    extra_inner_launch_call_args = ', '.join(internal_args_list)
+    macro_defs += gen_c_def_macro("EXTRA_INNER_LAUNCH_CALL_ARGS",
+                                  ', ' + extra_inner_launch_call_args if extra_inner_launch_call_args else "")
+    src = macro_defs + Path(os.path.join(dirname, "cuda_launcher.c")).read_text()
```

Comment:
personally I find this harder to read/modify than the previous version. 
Can we only move the functions that don't need patching in a separate file and keep the rest in a template kind of way (and maybe break up functions if it helps moving more code out of python)

Reply:
I get your point - let me update